File: src/udemy_tutorials_cpp/src/robot_news_station.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"


class RobotNewsStationNode: public rclcpp::Node
{
    public:
        RobotNewsStationNode(): Node("robot_news_station"), robot_name_("R2D2")
        {
            // Create a publisher for the robot_news topic
            publisher_ = this->create_publisher<example_interfaces::msg::String>("robot_news", 10);
            timer_ = this->create_wall_timer(
                std::chrono::seconds(1),
                std::bind(&RobotNewsStationNode::publish_news, this)
            );
            RCLCPP_INFO(
                this->get_logger(), 
                "Robot News Station has been started"
            );
        }

    private:
        void publish_news()
        {
            // Create a message
            auto msg = example_interfaces::msg::String();
            msg.data = 
                std::string("Hi, This is ") + 
                robot_name_ + 
                std::string(" from the Robot News Station");
            publisher_->publish(msg);
        }
        std::string robot_name_;
        rclcpp::Publisher<example_interfaces::msg::String>::SharedPtr publisher_;
        rclcpp::TimerBase::SharedPtr timer_;
};


int main(int argc, char **argv)
{
    // Initialize ROS2
    rclcpp::init(argc, argv);
    // Create a RobotNewsStationNode object
    auto node = std::make_shared<RobotNewsStationNode>();
    // Spin the node
    rclcpp::spin(node);  // This is like a while true loop
    // Shut down ROS2 node
    rclcpp::shutdown();
    return 0;
}

File: src/udemy_tutorials_py/udemy_tutorials_py/robot_news_station.py
#!/usr/bin/env python3
"""Example of using ROS2 publisher to a topic with Python."""
import sys
import rclpy
from rclpy.node import Node
from example_interfaces.msg import String


class RobotNewsStationNode(Node):
    """Node class where Robot News are published."""
    def __init__(self):
        # It is always recommended to call the node as the file name
        super().__init__("robot_news_station")
        # Create publisher
        self.__publisher = self.create_publisher(
            msg_type=String,
            topic="robot_news",
            qos_profile=10,  # Queue size. Normally 10 is enough.
        )
        # We are going to create a Timer to publish news
        self.create_timer(
            timer_period_sec=0.5,  # In seconds
            callback=self.publish_news,
        )
        # Show message that th node is alive
        self.logger.info("Robot News Station has been started.")

    @property
    def logger(self):
        """Get logger."""
        return self.get_logger()

    def publish_news(self):
        """Publish news to the topic."""
        # Create message
        msg = String()
        msg.data = "Hi, this is Robot News Station!"
        # Publish message
        self.__publisher.publish(msg)


def main(args=None):
    """Main function."""
    rclpy.init(args=args)
    node = RobotNewsStationNode()
    rclpy.spin(node)
    rclpy.shutdown()
    return 0


if __name__ == "__main__":
    sys.exit(main())
